### PR TITLE
added ping/pong set for 30s for each socket for #GH-1410

### PR DESCRIPTION
**Pull request by Zate**
This should disconnect sockets if no response to "ping" in 30s.  I've tested it locally with Linux, it should be tested through traefik and other proxies to ensure it works as it is supposed to.

Signed-off-by: Zate <zate75@gmail.com>

### CODE REVIEW

**Comment by marcdumais-work** [comment]
I have given this version a try, with traefik, and it did not seem to work as intended. (i.e. I got the disconnected state for terminals, navigator, editors,...) 

@Zate could you please give me pointers about how to capture / see the ping-pong packets using tcpdump? I will try some troubleshooting tomorrow.

**Comment by Zate** [comment]
To see the ping/pong on tcpdump I was simply tcpdumping for port 3000 on the host I had run the example/browser startup command on.  I was also using chrome://net-internals/#events&q=type:SOCKET%20is:active to see the ping/pong.  You wont see the ascii for ping pong, just regular connections.

File: packages/core/src/node/messaging/connection.ts [hunk 1]
```diff
@@ -40,10 +44,37 @@ export interface OnOpen {
...
+        if (extWs.isAlive === false) {
+            ws.terminate();
+            return;
+        }
+
+        extWs.isAlive = false;
+        ws.ping(null, undefined);
+
+    }, 30000);
+
```
Comment (marcdumais-work) [inline packages/core/src/node/messaging/connection.ts]: As mentioned above, wss does not emit "connection" when the web socket is created and opened.  So the listener above will never trigger. A way around that is to emit it ourselves. To do so we have to change the code below a bit. Something like:

```typescript
if (webSocket.readyState === webSocket.OPEN) {
   onOpen(webSocket, request, socket, head);
   wss.emit('connection', webSocket);
} else {
   webSocket.on('open', () => {
      onOpen(webSocket, request, socket, head);
      wss.emit('connection', webSocket);
   });
}
```

**Comment by marcdumais-work** [comment]
@Zate You are very close to this working. See review comments. Thanks again!

**Comment by marcdumais-work** [comment]
Hi @Zate . We will have a Theia release next week, and I would really like to revolve the disconnection issue. If you think you can answer the comments relatively quickly, then it's fine. Else I can help by creating a new PR, that would include your commit above and I would add an additional commit where I make the suggested corrections. 

Please let me know what you think.

**Comment by Zate** [comment]
Nice feedback!  I will take a look at this on Friday night and hopefully have a PR in over the weekend for it.  Thanks again for the feedback.

File: packages/core/src/node/messaging/connection.ts [hunk 2]
```diff
@@ -40,10 +44,37 @@ export interface OnOpen {
 }
 
 export function openSocket(options: IServerOptions, onOpen: OnOpen): void {
+
     const wss = new ws.Server({
         noServer: true,
         perMessageDeflate: false
     });
+
+    wss.on('connection', (ws: ws) => {
```
[the remark this answers is not in the document]
Comment (Zate) [reply packages/core/src/node/messaging/connection.ts]: ok will do.  I use vscode also but I don't think my setup works very well for javascript, it's setup for golang.  I'll see what I can do here.

File: packages/core/src/node/messaging/connection.ts (hunk 1 above)
Comment (Zate) [reply packages/core/src/node/messaging/connection.ts]: excellent I kind of felt like it needed something in that area but I was not sure what, I will get this in and tested.

File: packages/core/src/node/messaging/connection.ts [hunk 3]
```diff
@@ -40,10 +44,37 @@ export interface OnOpen {
...
+        const extWs = ws as ExtWebSocket;
+        extWs.isAlive = true;
+
+        extWs.on('pong', () => {
+            extWs.isAlive = true;
+        });
+
+    });
+
+    setInterval((ws: ExtWebSocket) => {
```
[the remark this answers is not in the document]
Comment (Zate) [reply packages/core/src/node/messaging/connection.ts]: Makes sense, thanks!

**Comment by marcdumais-work** [comment]
Hi @Zate 

You seem to have has an issue with rebasing your commit, I guess to the latest master. I can suggest a way to get back on your feet:

- checkout your work branch
```git checkout GH-1410```
- reset the branch to the commit reviewed above: 
```git reset --hard 9f97528dab0b849c4172b28328aa62ea295780a2```
- rebase your branch to the latest master (note: replace ```upstream``` if needed by the remote name you have for the main Theia repository)
```git rebase upstream/master```

or

- alternatively, if your local master branch is already in sync with the main Theia repository :
```git rebase master```

Normally, in this case, the rebase should be trivial, with no conflict to deal with. 
- force push to the PR branch: 
```git push -f origin GH-1410```

Let me know if you need more help

**Comment by marcdumais-work** [comment]
Also: once you have done the above, there should be no need to rebase your changes. In fact you can even skip the rebase step from above.

As long as nothing committed on master conflicts with your changes, GitHub will take care of the rebase for you when we push the PR. As long as you have the following message, that merging can be done automatically, you're good: 


image

**Comment by Zate** [comment]
Got it, was trying with that last night. I've made the changes and tested them to work well, I'll try and get the rebase/commit done today.

**Comment by Zate** [comment]
Ok I think i got everything done correctly.  It's been tested on my local and seems to work well.  Let me know what else I might need to do.

**Comment by marcdumais-work** [comment]
@Zate The code looks good.  I did a quick local test and it seems to works as intended - thanks! I will test it tomorrow at work with traefik just to confirm.

Could you please:
- Squash the two commits together into one (if I took care of this update I would have preserved your original commit and added a second one on top so that the trace of your contribution would not be lost. No point keeping it as two commits ATM). [how to squash](https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit)
- Alternatively, in a simple case like this, with a single file changed, you could take a shortcut, instead of squashing: 
  - save a copy of the modified file somewhere
  - reset your branch to the commit just before your first commit in this PR: ```git reset --hard c751dde4c44a257ebe3e501379930cb31d429106```
  - copy the modified file over the one that's there now
  - add/commit 

- Improve the commit message. The current commit message subject is longer than convention (~50 chars). When everything doesn't fit in a short string, it's best to have separate subject and body. For the subject, use something like: "Add ping/pong for maintaining WS through proxies". For the body, you could add a mention that this fixes socket disconnections when using a proxy such as traefik. It's also good to add in the body a reference to the GH issue: "fixes #1410". The way to separate subject from body is to leave an empty line between. e.g. : 
```
git commit -s -m 'commit msg subject
> 
> this is the commit message body'
```
result:
```
commit 627570a68877e87b7d3837329712a767f2d63a68
Author: Marc Dumais <marc.dumais@ericsson.com>
Date:   Sun Mar 25 07:16:12 2018 -0400

    commit msg subject
    
    this is the commit message body
    
    Signed-off-by: Marc Dumais <marc.dumais@ericsson.com>
```
Thanks again!

**Comment by Zate** [comment]
thanks for all the help Marc.  I think it's squashed.  It shows 2 commits but I think that is because I did an amend as the signature wasn't the same.  DCO shows green now.

**Comment by marcdumais-work** [comment]
Hi @Zate . I have just tested your fix in an environment that has a proxy (traefik) and it does work as we hoped. Thanks again for your patience.

Unless there are objections, I will be merging this PR in a few hours.

**Pull request merged**